### PR TITLE
HCPE-830 - Add TGW attachment resource

### PR DESCRIPTION
**Pull request by roaks3**
Ticket: https://hashicorp.atlassian.net/browse/HCPE-830

Adds the `hcp_aws_transit_gateway_attachment` resource and data source. The implementation is largely pulled from the initial implementation: https://github.com/hashicorp/cloud-terraform-provider-hcp-internal/pull/8

The inclusion of `aws` in the resource name came from some discussion captured [here](https://hashicorp.atlassian.net/browse/HCPE-872), which might change moving forward.

A simple example of what the resource looks like:
```
resource "hcp_aws_transit_gateway_attachment" "example" {
  hvn_id                        = hcp_hvn.main.hvn_id
  transit_gateway_attachment_id = "example-tgw-attachment"
  transit_gateway_id            = aws_ec2_transit_gateway.example.id
  resource_share_arn            = aws_ram_resource_share.example.arn
  destination_cidrs             = [aws_vpc.example.cidr_block]
}
```

### To Test

The example included here can be used to test the functionality of this resource. Note that the AWS config cannot be the same account as the HVN.

### CODE REVIEW

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (roaks3) [inline docs/data-sources/aws_transit_gateway_attachment.md:1]: The `docs/*` files included here were generated with `go generate`, which pulls in the examples and resource/field descriptions.

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (roaks3) [inline docs/data-sources/aws_transit_gateway_attachment.md:31]: The `state` field used here was copied from the internal implementation, but I think it should change to a `wait_for_active_state` flag, or `desired_state`. More detail below:

For this use case, I found some prior discussion from the existing AWS and Google providers:
- https://github.com/hashicorp/terraform-provider-aws/pull/3485
- https://github.com/hashicorp/terraform-provider-google/pull/4797

In short, there seems to be two concerns with using the field `state`:
1. It is merged with the computed `state` of the resource, which can make things behave strangely when trying to resolve whether the `state` has been set by the user or the service ([ref](https://github.com/hashicorp/terraform-provider-google/pull/4797#pullrequestreview-353254440)).
2. It is ambiguous whether the field is meant to wait for the specified state, or filter by the specified state. The [AWS provider is already using state fields as filters](https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/data_source_aws_vpc_peering_connection.go#L120-L130), so it could be confusing if users expect that same usage here.

I came across these two general solutions to the problem that I think we should consider:
1. `wait_for_active_state = "true"` ([ref](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecs_service#wait_for_steady_state))
2. `desired_state = "ACTIVE"` ([ref](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance#desired_status))

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (roaks3) [inline internal/provider/resource_aws_transit_gateway_attachment.go:66]: Nit: Our internal implementation used the `destination_cidrs` name that was copied here, but I haven't seen the "destination" qualifier anywhere else. Would `cidrs` be more concise/clear?

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (roaks3) [inline internal/provider/resource_aws_transit_gateway_attachment.go:194]: We will want to move this block up above the `WaitForOperation` call, but holding off on changing this for now because https://github.com/hashicorp/terraform-provider-hcp/pull/59 is still in-progress.

Comment (alexmunda) [reply internal/provider/resource_aws_transit_gateway_attachment.go:191]: https://github.com/hashicorp/terraform-provider-hcp/pull/59 has been merged. +1 to moving the id setting above the operation polling

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (bcmdarroch) [reply docs/data-sources/aws_transit_gateway_attachment.md:31]: Iiiinteresting conundrum 🤔 I'm leaning towards #1, unless we can think of any other state the user might want to wait for (maybe PENDING ACCEPTANCE). I've got some questions though.

> It is merged with the computed `state` of the resource
Do you mean terraform a different state that TF is tracking?

Say we go with #1. What happens if the resource never gets set to active? How can we find out its state?

Comment (xargs-P) [reply docs/data-sources/aws_transit_gateway_attachment.md:31]: What are the current different states of a TGW connection? 
Also I assume this is different from the state discussion over in: https://github.com/hashicorp/terraform-provider-hcp/pull/59

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (bcmdarroch) [reply internal/provider/resource_aws_transit_gateway_attachment.go:63]: I like the concise `cidrs`, assuming the AWS provider uses the same. Though I'd defer to @xargs-P

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (bcmdarroch) [reply internal/provider/resource_aws_transit_gateway_attachment.go:191]: Does that need to happen in any other resource or is this the last one?

File: internal/provider/resource_hvn.go
Comment (bcmdarroch) [inline internal/provider/resource_hvn.go:92]: Let's save this for a follow-up, but I think this is a good point to introduce the pattern for returning cloud-specific outputs.  I'm thinking we'll have an output with the top-level name `aws`, which contains `account_id`, similar to this [vpc_config object](https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_eks_cluster.go#L165-L214) in the the AWS provider's eks_cluster resource. It's a bit of a hack, but it sounds like it's the best we've got while the provider SDK team works out a better solution for input/output objects.

**Review by bcmdarroch — APPROVED** [review]
Had some questions and a suggestion for the new `provider_account_id` HVN output, but otherwise nice work! ✨

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (alexmunda) [reply docs/data-sources/aws_transit_gateway_attachment.md:31]: > Iiiinteresting conundrum 🤔 I'm leaning towards #1, unless we can think of any other state the user might want to wait for (maybe PENDING ACCEPTANCE). I've got some questions though.
> 
> > It is merged with the computed `state` of the resource
> > Do you mean terraform a different state that TF is tracking?
> 
> Say we go with #1. What happens if the resource never gets set to active? How can we find out its state?

Our polling logic should account for any "terminal" state (errored, etc that we can never recover from) and throw an error. 

I prefer the `wait_for_active_state` approach of option 1 as it puts less burden on the user to inspect every possible state. It would also align with doing the same for peering connections as discussed in https://hashicorp.atlassian.net/browse/HCPE-871

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (alexmunda) [reply internal/provider/resource_aws_transit_gateway_attachment.go:191]: TGW should be the only resource right now that doesn't set the `id` before polling.

File: internal/clients/tgw.go
Comment (alexmunda) [inline internal/clients/tgw.go:42]: Are there any other terminal states to be aware of in regards to TGW?

File: internal/clients/tgw.go
Comment (alexmunda) [inline internal/clients/tgw.go:37]: Does `Gateway` need to be capitalized in these error messages?

File: internal/provider/data_source_aws_transit_gateway_attachment.go
Comment (alexmunda) [inline internal/provider/data_source_aws_transit_gateway_attachment.go:36]: I think if we keep this approach, we need to list out the possible states that the user can pick and validate them on the input.

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (alexmunda) [inline internal/provider/resource_aws_transit_gateway_attachment.go:60]: How come the ARN is sensitive in this case? I usually think of AWS ARNs synonymous to unique IDs, have I been lying to myself this whole time!?

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (alexmunda) [inline internal/provider/resource_aws_transit_gateway_attachment.go:141]:
```suggestion
	log.Printf("[INFO] HVN (%s) found, proceeding with TGW create", hvnID)
```

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (roaks3) [reply docs/data-sources/aws_transit_gateway_attachment.md:31]: For reference, possible states for TGW attachment currently are ([ref](https://github.com/hashicorp/cloud-network/blob/master/proto-public/20200907/tgw_attachment.proto#L47-L55)):
```
UNSET
CREATING
PENDING_ACCEPTANCE
ACCEPTED
ACTIVE
FAILED
REJECTED
DELETING
EXPIRED
```

| unless we can think of any other state the user might want to wait for (maybe PENDING ACCEPTANCE)
Yea, it seems like `PENDING_ACCEPTANCE` and `ACTIVE` would be the two states we would be interested in

| Do you mean terraform a different state that TF is tracking?
There's more context in [this link](https://github.com/hashicorp/terraform-provider-google/pull/4797#pullrequestreview-353254440), but I think the concern is that the `state` could be updated from the server during a READ operation, and then you don't really know anymore what state the user is waiting for.

| Say we go with #1. What happens if the resource never gets set to active? How can we find out its state?
Yea, what @alexmunda mentioned sounds like the right solution in that case: store the state of the resource as it gets updated within the operation.

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (roaks3) [reply internal/provider/resource_aws_transit_gateway_attachment.go:191]: Yep, it looked like Alex got all of the other cases of this in his PR. Will be updating this one to match

File: internal/provider/resource_hvn.go
Comment (roaks3) [reply internal/provider/resource_hvn.go:92]: Yea, I think this would warrant some more discussion, so my preference would be to include that in the general topic around cloud-specific values in our network resources. Definitely agree that this field falls into that category.

File: internal/clients/tgw.go
Comment (roaks3) [reply internal/clients/tgw.go:42]: I'm not 100% sure, but I can check and make sure we're returning once we hit any terminal state. My guess is that we need to include a few more here.

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (smaant) [reply internal/provider/resource_aws_transit_gateway_attachment.go:63]: this is essentially the same what AWS calls `destination-cidr-block` https://docs.aws.amazon.com/cli/latest/reference/ec2/create-route.html. Relatively soon we will introduce a concept of HVN Routes that would have destination-cidr (ip range) and destination (peering, tgw attachment, etc). I guess we can call it just `cidr`. In the peering, for example, it's called `peer_vpc_cidr_block`.

File: internal/clients/tgw.go
Comment (roaks3) [reply internal/clients/tgw.go:37]: I'm honestly not sure here. Our capitalization guide from design doesn't seem to explicitly call out a case like this: https://docs.google.com/document/d/1ABNkYIp6aLwPYNpuYOFcDFBFCcnH_4ePRaLP-gD_tKE/edit

@joshklekamp Do you have an opinion here on how we capitalize `Transit gateway attachment` within our error messages and logs?

**Review by smaant — COMMENTED** [review]
A few comments

File: internal/provider/data_source_aws_transit_gateway_attachment.go
Comment (roaks3) [reply internal/provider/data_source_aws_transit_gateway_attachment.go:36]: Yea, I was waiting to see where we landed with the interface, but will make sure to address documentation, validation, and diffing once we make a choice.

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (smaant) [reply internal/provider/resource_aws_transit_gateway_attachment.go:141]: "TGW attachment"

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (roaks3) [reply internal/provider/resource_aws_transit_gateway_attachment.go:60]: Hmm, this came from @smaant , but I haven't actually verified. Anton, do you know if this is a sensitive value?

File: internal/clients/tgw.go
Comment (joshklekamp) [reply internal/clients/tgw.go:37]: It should be sentence case unless we add AWS to the beginning.

Comment (joshklekamp) [reply internal/clients/tgw.go:37]: or Amazon 😄

File: docs/resources/aws_transit_gateway_attachment.md
Comment (nombiezinja) [reply docs/resources/aws_transit_gateway_attachment.md:76]: Perhaps it makes sense to be explicit about it and say "The user-settable name of the Transit Gateway attachment in HCP"

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (smaant) [reply internal/provider/resource_aws_transit_gateway_attachment.go:60]: in general ARNs aren't sensitive but we kind of treat them as such. The idea is that Resource Share (RS) is shared with the entire HVN's org AWS account, therefore technically any HVN in this org can use this RS to attach to TGW shared in this RS as long as they know TGW ID and RS ARN therefore we need to keep one them as a secret.

File: internal/clients/tgw.go
Comment (smaant) [reply internal/clients/tgw.go:37]: AWS itself calls it "Transit Gateway"

File: internal/clients/tgw.go
Comment (smaant) [reply internal/clients/tgw.go:42]: technically there're also `REJECTED` and `EXPIRED`, we might add it here too

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (nombiezinja) [inline internal/provider/resource_aws_transit_gateway_attachment.go:23]:
```suggestion
		Description: "The AWS Transit gateway attachment resource allows you to manage a Transit Gateway attachment that attaches an HVN to a Transit Gateway you own in AWS.",
```

File: internal/clients/tgw.go
Comment (nombiezinja) [reply internal/clients/tgw.go:42]: Other terminal states in addition to `FAILED` include `REJECTED`, `DELETING`, `EXPIRED`. 
`UNSET` being returned should also indicate something very wrong as well, however this should never happen, we are probably ok not handling this case here.

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (nombiezinja) [inline internal/provider/resource_aws_transit_gateway_attachment.go:152]: Hmm not sure if this is the pattern we follow everywhere so please ignore this comment if this is a custom thing, but [a check ensuring unique tgwAttachmentID within a project is in place](https://github.com/hashicorp/cloud-network/blob/master/service/network/private/service_create_tgw_attachment.go#L341), and would be part of the output in the subsequent `Create` request.

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (xargs-P) [reply docs/data-sources/aws_transit_gateway_attachment.md:31]: Thanks for the comments everyone. This is a good thread. I agree with the teams viewpoint on the 1st item.

File: docs/resources/aws_transit_gateway_attachment.md
Comment (xargs-P) [reply docs/resources/aws_transit_gateway_attachment.md:10]: @smaant 
>  to specify that a Transit Gateway is owned by a customer (located in their account) 
++ 
> should be located in the same AWS region as the HVN 
I don't think we have any technically limitation stopping users from having the connection cross regions. How about we change the text to "....and _for a performant connection_ should be located in the same AWS region as the HVN" ?

File: docs/resources/aws_transit_gateway_attachment.md
Comment (xargs-P) [inline docs/resources/aws_transit_gateway_attachment.md:82]: 1. Do we have any other resources that have nested timeouts? 🤔  
2. Can we add this to the example above?

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (smaant) [reply docs/data-sources/aws_transit_gateway_attachment.md:31]: In case of going with `wait_for_active_state` approach I would recommend deleting `state` input, otherwise you'll have to deal with edge-cases when `wait_for_active_state = true` and `state` is set to something different.

File: docs/resources/aws_transit_gateway_attachment.md
Comment (smaant) [reply docs/resources/aws_transit_gateway_attachment.md:10]: > I don't think we have any technically limitation stopping users from having the connection cross regions.

@xargs-P we do, AWS doesn't allow cross region TGW VPC attachments. If customer wants to have their TGW and HVN in different regions they will need to create a new TGW in the same region as HVN and peer these TGWs. But TGW attachment from HVN have to be to a TGW in the same region.

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (roaks3) [reply docs/data-sources/aws_transit_gateway_attachment.md:31]: Agreed, thanks for the comments! I'll proceed with a `wait_for_active_state` flag and change `state` to be read-only.

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (roaks3) [reply docs/data-sources/aws_transit_gateway_attachment.md:30]: So this `id` represents our unique Terraform id stored in the state (format is `/project/123/hashicorp.network.tgw-attachment/456`), but the description here is not particularly descriptive, and I think it might actually be confusing for users. Note that it exists here because it is autogenerated by `go generate`.

I think we should leave it for now, because it aligns with the generator and our other resources, but I've created a ticket to see if it makes sense to modify our docs generation to not include these: https://hashicorp.atlassian.net/browse/HCPE-883

File: docs/resources/aws_transit_gateway_attachment.md
Comment (roaks3) [reply docs/resources/aws_transit_gateway_attachment.md:82]: These are actually autogenerated by `go generate`, and allow the user to specify custom timeouts for their operations. We do have these on all of our other resources/data sources currently, so the changes here align with that.

We could include these in the example code, but I think it could be a bit distracting for a user? I would expect a user to only need to set these timeouts in very rare situations, if at all. In the case that these are not useful, we may want to consider just removing them from the documentation (similar to the `id` https://hashicorp.atlassian.net/browse/HCPE-883)

File: docs/resources/aws_transit_gateway_attachment.md
Comment (roaks3) [reply docs/resources/aws_transit_gateway_attachment.md:98]: These fields, and the `default` field in the data source, are meant for setting custom timeouts for the operations. They render [like this](https://registry.terraform.io/providers/hashicorp/hcp/latest/docs/resources/hvn#nested-schema-for-timeouts) on the docs site, and are currently documented this way for all of our resources due to `go generate`.

Similar to my comment on the `id`, I think we should leave these they way they are for consistency, but make some sort of adjustment moving forward so that they are more clear to users. Possible options include removing these blocks from the docs, or adding actual descriptions to explain what these fields do.

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (roaks3) [reply internal/provider/resource_aws_transit_gateway_attachment.go:152]: Ah yea, this came up with the HVN resource, and we ended up with the pattern you see here where we do a GET first: https://github.com/hashicorp/terraform-provider-hcp/pull/3#discussion_r556234766

We might be able to accomplish the same thing with the CREATE call, but I think we would need to be careful about what guarantees we have from the endpoint (eg. it looks like we can get at least two types of `AlreadyExists` response codes), and the main driver here was being able to supply an actionable error message.

Comment (nombiezinja) [reply internal/provider/resource_aws_transit_gateway_attachment.go:152]: Thanks for explaining, this sounds good to me, resolving this comment

File: internal/clients/tgw.go
Comment (roaks3) [reply internal/clients/tgw.go]: Thanks for the additional context here (the diagram was really helpful too). I've updated the function to actually specify the "pending" states, instead of terminal states, to align with [the helper that the provider SDK offers](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/state.go). This helper is used frequently in other providers, and seems to be a safer approach if we introduce new states in the future.

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (smaant) [inline docs/data-sources/aws_transit_gateway_attachment.md:43]: `state` is missing

File: internal/clients/tgw.go
Comment (roaks3) [reply internal/clients/tgw.go]: Verified with Josh that we should use sentence case here (so `Transit gateway attachment` at the beginning of a sentence, but `transit gateway attachment` in the middle of a sentence). There are two exceptions where we refer to the resource/data source by the full name in their top-level description, and in those cases we use `AWS Transit Gateway Attachment`. I've updated the error messages and descriptions to reflect the correct capitalization.

File: docs/data-sources/aws_transit_gateway_attachment.md
Comment (roaks3) [reply docs/data-sources/aws_transit_gateway_attachment.md]: Resolving based on conversation here: https://github.com/hashicorp/terraform-provider-hcp/pull/58#discussion_r574561134

File: docs/resources/aws_transit_gateway_attachment.md
Comment (roaks3) [reply docs/resources/aws_transit_gateway_attachment.md:82]: Created https://hashicorp.atlassian.net/browse/HCPE-890 to address these `timeouts` fields across all resources

File: internal/provider/resource_aws_transit_gateway_attachment.go
Comment (roaks3) [reply internal/provider/resource_aws_transit_gateway_attachment.go:63]: Given @smaant 's response, I'm going to leave this as `destination_cidrs`. Also note that the UI field is labeled "Destinations", so I think this aligns well with that.

**Review by smaant — APPROVED** [review]

**Comment by smaant** [comment]
Great job @roaks3, thank you for handling this!

**Review by nombiezinja — APPROVED** [review]
Awesome work, appreciate you taking this on!

**Pull request merged**
